distributed/scheduler.py

```diff
@@ -4068,6 +4068,28 @@ def decide_worker(self, ts):
 
         return worker
 
+    def set_duration_estimate(self, ts, ws):
+        """Estimate task duration using worker state and task state.
+
+        If a task takes longer than twice the current average duration we
+        estimate the task duration to be 2x current-runtime, otherwise we set it
+        to be the average duration.
+        """
+        # we might want to pull these vars from settings?
+        support_outliers = True
+        threshold = 2
+
+        duration = self.get_task_duration(ts)
+        comm = self.get_comm_cost(ts, ws)
+        new_duration = duration + comm
+        exec_time = -1
+        if 'executing' in ws.metrics and ts.key in ws.metrics['executing']:
+            exec_time = ws.metrics['executing'][ts.key]
+        if support_outliers and exec_time > threshold * duration:
+            new_duration = 2 * exec_time
+        ws.processing[ts] = new_duration 
+        return ws.processing[ts]
```

Comment:
Some thoughts here:

-  I wonder if we should place `executing` on the WorkerState class directly, rather than including it in the catch-all metrics attribute.  
-  Until we decide to generalize things, I'd prefer to assume values of `support_outliers` and `threshold`.  I don't think that we should generalize pre-maturely.  I suspect that by inlining these this code will become a bit simpler and so be easier for future maintainers to quickly understand.

Thoughts?

Reply:
For `executing`, I agree it makes sense though I'm not familiar enough with that code to speak meaningfully about it.

For the second point, I had considered the possibility that those parameters would become configuration options so users could tweak them for weird cases.  I was initially thinking of the possibly of fat gaussians (where the variance is larger than the mean) but then I realized that can't happen with purely positive data.  Given that, its probably fine to hard-code default values.

I still think the having a separate function for this makes sense though since the duration is computed in two places.  It will also make it easier to change the logic later if needed and I would think the function name makes the code more navigable.

Reply:
> For executing, I agree it makes sense though I'm not familiar enough with that code to speak meaningfully about it.

The relevant code on the scheduler side is here:

https://github.com/dask/distributed/blob/cee4e3c99c34c1e515b9e80032d19c24b5f390ed/distributed/scheduler.py#L1614-L1623

We might add an `executing={}` dict to that signature

The relevant code on the worker side is in two places

https://github.com/dask/distributed/blob/cee4e3c99c34c1e515b9e80032d19c24b5f390ed/distributed/worker.py#L911-L916

https://github.com/dask/distributed/blob/cee4e3c99c34c1e515b9e80032d19c24b5f390ed/distributed/worker.py#L785-L788

I would remove the executing bit from get_metrics, and send it as a direct parameter to `heartbeat_worker`.  I would then modify the WorkerState object on the Scheduler directly.

This could be a separate PR, or we could just roll it into this one if that's simpler (personally I would probably just roll it in).

Reply:
> For the second point, I had considered the possibility that those parameters would become configuration options so users could tweak them for weird cases. I was initially thinking of the possibly of fat gaussians (where the variance is larger than the mean) but then I realized that can't happen with purely positive data. Given that, its probably fine to hard-code default values.

Yeah, I think that these are valid points, and something that we might want to do in the future.  In general though there are a *lot* of different things that we've said we might want to do and never did.  If we generalized things when possible like this then I think that the codebase would quickly become unreadable.  Something I've learned after a while of doing this is to optimize strongly for future reviewer readability.  If we can make this a five-line function with only one conditional at the cost of generality then I think that that is a clear win.

Reply:
+1 for holding off on generalizing

Reply:
Ok sounds good but do you agree that it makes sense to keep it a separate function rather than inline?

Reply:
Separate function seems fine to me